src/python/pants/backend/python/typecheck/mypy/rules.py

```diff
@@ -273,28 +289,21 @@ async def mypy_typecheck_partition(
                             # for different versions) and uses a one-process-at-a-time daemon by default,
                             # multiple MyPy processes operating on a single db cache should be rare.
 
-                            {mkdir.path} -p {run_cache_dir}/{py_version} > /dev/null 2>&1 || true
-                            {cp.path} {mypy_cache_dir}/{py_version}/cache.db {run_cache_dir}/{py_version}/cache.db > /dev/null 2>&1 || true
+                            NAMED_CACHE_DIR="{mypy_cache_dir}/{py_version}"
+                            NAMED_CACHE_DB="$NAMED_CACHE_DIR/cache.db"
+                            SANDBOX_CACHE_DIR="{run_cache_dir}/{py_version}"
+                            SANDBOX_CACHE_DB="$SANDBOX_CACHE_DIR/cache.db"
+
+                            {mkdir.path} -p "$SANDBOX_CACHE_DIR" > /dev/null 2>&1
+                            {cp.path} "$NAMED_CACHE_DB" "$SANDBOX_CACHE_DB" > /dev/null 2>&1
```

Comment:
One more annoying concurrency question about this existing code... if a another process is concurrently mutating this SQLite DB, does this cp do the right thing? For example, could we end up with a "torn read" where we copy half the before-mutation DB and half the after-mutation DB and end up with a corrupted file?

I guess this depends on cp and/or SQLite?

Reply:
If I'm following what you're saying, then no, I don't think so.

1. We always copy from the named cache into the sandbox before we start the process
2. Then we run mypy to completion
3. Then we atomically replace the named cache value

Mypy is only ever operating on the copy.

You could ask if the copy will ever have its data changed file it's copying, but because of the atomic operations, that's also a no. Once it opens the file descriptor, that should be unchanged for the entirety of the copy.

Reply:
Ah yeah, I see. You understood what I was asking.

Reflecting back to you to close the loop: if mypy ever runs directly on the cache, it's not pants doing it, and "user doing silly games in the pants cache dirs" is out of scope for pants to handle (and, as we saw with pyenv, marking things read-only to reduce risk doesn't seem to work well).

Thanks!

Reply:
Yeah. To me this whole ordeal is a stopgap until we find a more performant way of handling mypy. Either through it's daemon or incremental typechecking.